File: playwright-e2e/pages/NewTransactionPage.ts
import { expect, type Locator, type Page } from "@playwright/test";

/** The three-step "new transaction" wizard at /transaction/new. */
export class NewTransactionPage {
  readonly search: Locator;
  readonly form: Locator;
  readonly amount: Locator;
  readonly description: Locator;
  readonly submitPayment: Locator;
  readonly submitRequest: Locator;
  readonly returnToTransactions: Locator;
  readonly createAnother: Locator;

  constructor(private readonly page: Page) {
    // Exception to the wrapper rule: the app passes this one through
    // `inputProps`, so the attribute is already on the <input>.
    this.search = page.getByTestId("user-list-search-input");
    this.form = page.getByTestId("transaction-create-form");
    this.amount = page.getByTestId("transaction-create-amount-input").locator("input");
    this.description = page.getByTestId("transaction-create-description-input").locator("input");
    this.submitPayment = page.getByTestId("transaction-create-submit-payment");
    this.submitRequest = page.getByTestId("transaction-create-submit-request");
    this.returnToTransactions = page.getByTestId("new-transaction-return-to-transactions");
    this.createAnother = page.getByTestId("new-transaction-create-another-transaction");
  }

  async goto(): Promise<void> {
    await this.page.goto("/transaction/new");
    await expect(this.search).toBeVisible();
  }

  /** Step 1 — the search is debounced, so wait for the specific row. */
  async selectContact(userId: string, searchTerm: string): Promise<void> {
    await this.search.fill(searchTerm);
    await this.page.getByTestId(`user-list-item-${userId}`).click();
    await expect(this.form).toBeVisible();
  }

  /** Step 2 — amount in dollars, e.g. "25.50". Empty strings leave a field blank. */
  async enterDetails(amount: string, description: string): Promise<void> {
    await this.amount.fill(amount);
    await this.description.fill(description);
  }

  async pay(): Promise<void> {
    await this.submitPayment.click();
  }

  async request(): Promise<void> {
    await this.submitRequest.click();
  }

  async expectConfirmation(): Promise<void> {
    await expect(this.returnToTransactions).toBeVisible();
    await expect(this.createAnother).toBeVisible();
  }

  async backToFeeds(): Promise<void> {
    await this.returnToTransactions.click();
  }

  async expectSubmitDisabled(): Promise<void> {
    await expect(this.submitPayment).toBeDisabled();
    await expect(this.submitRequest).toBeDisabled();
  }
}

File: selenium-e2e/src/test/java/dev/dsbon/realworld/pages/NewTransactionPage.java
package dev.dsbon.realworld.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

/** The three-step "new transaction" wizard at /transaction/new. */
public class NewTransactionPage extends BasePage {

  // Exception to the wrapper rule: the app passes this one through `inputProps`,
  // so the attribute already sits on the <input>.
  private static final By SEARCH = testId("user-list-search-input");
  private static final By FORM = testId("transaction-create-form");
  private static final By AMOUNT = testIdInput("transaction-create-amount-input");
  private static final By DESCRIPTION = testIdInput("transaction-create-description-input");
  private static final By SUBMIT_PAYMENT = testId("transaction-create-submit-payment");
  private static final By SUBMIT_REQUEST = testId("transaction-create-submit-request");
  private static final By RETURN_TO_TRANSACTIONS = testId("new-transaction-return-to-transactions");
  private static final By CREATE_ANOTHER = testId("new-transaction-create-another-transaction");

  public NewTransactionPage(WebDriver driver) {
    super(driver);
  }

  public NewTransactionPage open() {
    navigateTo("/transaction/new");
    visible(SEARCH);
    return this;
  }

  /**
   * Step 1 — pick a contact.
   *
   * <p>The search is debounced, so the list re-renders after the last keystroke.
   * A single click dispatched into that window lands on a node React is about to
   * replace and is silently lost — no exception, no navigation, and then a
   * timeout on the form that follows. {@code clickUntil} verifies the real
   * outcome (the payment form appeared) and re-clicks until it does.
   */
  public NewTransactionPage selectContact(String userId, String searchTerm) {
    fill(SEARCH, searchTerm);
    By item = By.cssSelector("[data-test='user-list-item-%s']".formatted(userId));
    clickUntil(item, d -> isVisible(d, FORM));
    return this;
  }

  /** Step 2 — amount in dollars, e.g. "25.50". An empty string leaves a field blank. */
  public NewTransactionPage enterDetails(String amount, String description) {
    fill(AMOUNT, amount);
    fill(DESCRIPTION, description);
    return this;
  }

  public NewTransactionPage pay() {
    click(SUBMIT_PAYMENT);
    return this;
  }

  public NewTransactionPage request() {
    click(SUBMIT_REQUEST);
    return this;
  }

  public NewTransactionPage awaitConfirmation() {
    visible(RETURN_TO_TRANSACTIONS);
    visible(CREATE_ANOTHER);
    return this;
  }

  public void backToFeeds() {
    click(RETURN_TO_TRANSACTIONS);
  }

  public boolean isPaymentEnabled() {
    return isEnabled(SUBMIT_PAYMENT);
  }

  public boolean isRequestEnabled() {
    return isEnabled(SUBMIT_REQUEST);
  }

  /**
   * Waiting variants, for any state reached through an interaction.
   *
   * <p>Selenium assertions do not retry, so reading {@link #isPaymentEnabled}
   * immediately after typing reads the DOM before React has re-rendered.
   */
  public NewTransactionPage awaitSubmitEnabled() {
    awaitEnabled(SUBMIT_PAYMENT);
    awaitEnabled(SUBMIT_REQUEST);
    return this;
  }

  public NewTransactionPage awaitSubmitDisabled() {
    awaitDisabled(SUBMIT_PAYMENT);
    awaitDisabled(SUBMIT_REQUEST);
    return this;
  }

  /**
   * Touch the amount field and leave it empty.
   *
   * <p>Like every other form in this app the buttons are bound to
   * {@code disabled={!isValid || isSubmitting}} and Formik starts with
   * {@code isValid === true} — so a pristine form has ENABLED buttons.
   */
  public NewTransactionPage touchAndClearAmount() {
    fill(AMOUNT, "1");
    fill(AMOUNT, "");
    visible(DESCRIPTION).click();
    return this;
  }
}
